### PR TITLE
New PlayerInteractEvent for 1.9

### LINKED ISSUE #2591: [1.9] PlayerInteractionEvent DISCUSSION
Here'll be the discussion for the next iteration of player interaction events (mainly, right clicks. the left click block interact event still works)
MC interactions are bloody and ugly, so we should put some thought into making these as painless and easy to understand as possible

I have traced the full interaction logic here, if you want to understand how it all flows: https://gist.github.com/williewillus/c781d9824df3ae723bff

Mainly we need to find/determine
* Hook points
* Semantics upon cancellation (old one had all sorts of ugly stuff if you canceled on only one side, etc. etc.)
* If it's possible, a writeup of the proper values to return from the various interaction callbacks in item, block, entity, to get the desired behaviour. (e.g. "must return the same thing on both sides in this method or else you get desync!" and stuff like that)

### PR DESCRIPTION
**Pull request by williewillus**
Closes #2591 if taken.

Notes:
* All interaction events grouped and centralized into one file. Yes, that includes the entity ones. I figured I could move those as well since everything's getting ripped up from the roots right now (shhandcpwsaiditwasagoodidea)
* Static subclasses for each type of interaction: Entity Interact_At, Entity Interact, right click block, right click item, left click block. 
* Side effect of reorganizing: No more RIGHT_CLICK_AIR. The case where that happens will always be the same as RightClickItem. However, it currently doesn't catch the case where you right click with an empty hand. That can easily be added on the client, but the server doesn't seem to know when you do that anymore (the section of code where packet is sent is nullchecked by stack)
* Each static subclass has what it specifically needs to succeed, instead of lumping everything into one object type like it used to be
* GREATLY improved the documentation on each event to detail specifically when it fires and what happens when you change the result or cancel it
* Reapplied patches to PlayerInteractionManager (specifically, onItemUseFirst and doesSneakBypassBlockUse)
* Event handling on the client and server are no longer tied together. With the old event, sometimes canceling clientside would cause the event to not even fire serverside, which according to some modders was a big point of confusion when working with this event. Hopefully changing this + better documentation will alleviate that.

Please post feedback and comments, and check my logic to make sure it's right.
(Leaving all commits for ease of discussion, squash will happen if it gets taken)

### CODE REVIEW

**Comment by cpw** [comment]
I would add a side argument to the event: client or server.

**Comment by williewillus** [comment]
Oh and also, it seems the server is no longer aware of when the player right clicks with an empty hand.
Should I just fire that clientside (haven't implemented that yet) and let mods handle it?

**Comment by williewillus** [comment]
Added some tests, which all pass at the moment.
Every subevent tested, except right click block (the monster/really complex one)
Will update tonight with more

**Comment by williewillus** [comment]
Added more tests which seem to work, I think this pie is baking really well ;p

Things left to do (potentially?)
* A "right click air with empty hand" event fired on the client (server is not aware of this)
* Left click block fire clientside

**Comment by kashike** [comment]
So the server does not know when a player right clicks with an empty hand? How about left click with empty hand?
That seems like something the server should know about, to be honest.

**Comment by williewillus** [comment]
Doesn't appear it does.
Left clicking a block fires the LeftClickBlock and left clicking an entity attacks it, but left clicking empty space just calls swingArm which sends an animation packet to the server to swing the arm (in other words, not a good place to hook and fire an interact event)

**Comment by williewillus** [comment]
Left clickblock now fires clientside.

If people would start testing it (or telling me test cases) we can get this pinned down.
Or, just submit the cases you used for PIE in the past and I'll add some form of them to the main test.

**Comment by williewillus** [comment]
Anyone tested this a bit yet?
I'd like to get this into forge asap because Botania heavily relies on PlayerInteractEvents.
All general cases should work 100%, need help testing edge cases

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 1]
```diff
@@ -1,78 +1,298 @@
...
+    }
+
+    /**
+     * This event is fired on both sides whenever a player interacts with an entity.
+     * It has the local position (where on the entity) the player interacted at.
+     * Canceling the event clientside will cause processing to continue to EntityInteract (below),
+     * while canceling serverside will simply do no further processing.
+     */
+    @Cancelable
+    public static class EntityInteractSpecific extends PlayerInteractEvent {
```
Comment (kashike) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: The formatting for the inner classes varies (`{` on same line instead of new line, etc).

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 2]
```diff
@@ -1,78 +1,298 @@
...
+    private final BlockPos pos;
+    private final EnumFacing face;
+
+    private PlayerInteractEvent(EntityPlayer player, EnumHand hand, ItemStack stack, BlockPos pos, EnumFacing face)
+    {
+        super(Preconditions.checkNotNull(player, "Null player in PlayerInteractEvent!"));
+        this.hand = Preconditions.checkNotNull(hand, "Null hand in PlayerInteractEvent!");
+        this.stack = stack;
+        this.pos = Preconditions.checkNotNull(pos, "Null position in PlayerInteractEvent!");
+        this.face = face;
```
Comment (kashike) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: Is face nullable? If so, might be a good idea to mark the parameter, field, and accessor with `@Nullable`.

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 3]
```diff
@@ -1,78 +1,298 @@
...
+    private final EnumHand hand;
+    private final ItemStack stack;
+    private final BlockPos pos;
+    private final EnumFacing face;
+
+    private PlayerInteractEvent(EntityPlayer player, EnumHand hand, ItemStack stack, BlockPos pos, EnumFacing face)
+    {
+        super(Preconditions.checkNotNull(player, "Null player in PlayerInteractEvent!"));
+        this.hand = Preconditions.checkNotNull(hand, "Null hand in PlayerInteractEvent!");
+        this.stack = stack;
```
Comment (kashike) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: Is stack nullable? If so, might be a good idea to mark the parameter, field, and accessor with `@Nullable`.

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java (hunk 1 above)
Comment (williewillus) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: what's the accepted convention for forge inner/nested classes?

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java (hunk 3 above)
Comment (williewillus) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: I thought we didn't want those annotations in the codebase?

Comment (williewillus) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: (I personally think they just muddle everything up, but if it's the convention I'll go with it. None of the other events seem to have anything though)

Comment (kashike) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: Somewhat used (https://github.com/MinecraftForge/MinecraftForge/search?utf8=%E2%9C%93&q=nullable), but in my opinion it's best to use them to be clear - IDEs will get warnings when not checking for null on those accessors.

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java (hunk 1 above)
Comment (kashike) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: https://github.com/MinecraftForge/MinecraftForge/blob/master/src/main/java/net/minecraftforge/event/entity/player/PlayerEvent.java#L28

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java (hunk 3 above)
Comment (williewillus) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: Also all of this is internal, only the getters would need to be annotated.
The checkStates are just there for internal forge sanity checks

**Comment by williewillus** [comment]
Fixed formatting, tweaked  comments a bit, added two @Nullables

**Comment by williewillus** [comment]
Clarified comments
Added another test
Fixed a desync from a patch that I didn't apply on both sides.
Fixed bug in RightClickBlock

Anyone testing this? :/
I'd like to get it in really soon (my mods need it badly) and this should be tested by everyone

**Comment by Guichaguri** [comment]
It works pretty well, this new API is easier to use and understand
I also really need this for a mod. :+1:

I will test server-side and cancelling the events tomorrow

**Comment by williewillus** [comment]
Thanks a lot!

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 4]
```diff
@@ -1,88 +1,322 @@
...
-    private final Action action;
-    private final World world;
-    private final BlockPos pos;
-    private final EnumFacing face; // Can be null if unknown
-    private final Vec3d localPos; // Can be null if unknown
+    /**
+     * This event is fired on both sides whenever a player interacts with an entity.
+     * It has the local position (where on the entity) the player interacted at.
+     * Canceling the event clientside will cause processing to continue to EntityInteract (below),
+     * while canceling serverside will simply do no further processing.
```
Comment (gabizou) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: IIRC, interacting with entities didn't always include the targeted position of where the entity was interacted *at*. Did this change in 1.9 where it's always sent from the client?

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 5]
```diff
@@ -1,88 +1,322 @@
...
+     */
+    @Cancelable
+    public static class EntityInteractSpecific extends PlayerInteractEvent
```
Comment (gabizou) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: It would make sense if this extended `EntityInteract` and share the target, adding only the local position.

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 6]
```diff
@@ -1,88 +1,322 @@
...
-    public PlayerInteractEvent(EntityPlayer player, Action action, BlockPos pos, EnumFacing face, World world)
+        public Entity getTarget()
+        {
+            return target;
+        }
+    }
+
+    /**
+     * This event is fired on both sides after EntityInteractSpecific is fired above.
+     * If the above event is canceled, this event will still fire.
```
Comment (gabizou) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: Why would this event fire? What is the case where it fires, and what is it actually signifying that the previous event was cancelled? Is this more of a notification event instead of "oh Item.onItemUse is going to be called".

**Comment by williewillus** [comment]
@gabizou in response to all of your comments

"Interact At" and "Interact" are two COMPLETELY different method calls. Everything about them is separate. The second is the old entity interaction, the first was added in 1.8 to allow armorstands to detect where the player clicked them. Why didn't mojang just retrofit the old one? No Idea.

Because of this distinction, I chose not to bind them together with inheritance. An "Interact At" event does not have a "IS-A" relationship with "Interact", they're separate actions as far as vanilla is concerned.

If you look at the patch for Minecraft.java you see the invocations for them are made separately and in the patch for NetHandlerPlayServer they are also processed completely separately.

**Comment by gabizou** [comment]
@williewillus Right, that's what I was wondering about. I was under the assumption from the docs of the two events that the interact at would always be thrown with entities (not just armor stands now), and *then* as a pseudo post the interact entity event would be thrown.

**Comment by williewillus** [comment]
Yea, I'll clarify that

**Comment by williewillus** [comment]
Do you have ideas for a better name for the specific one?

**Comment by gabizou** [comment]
So, what probably needs to be clarified is:

- Are both events always going to be thrown for interacting at ArmorStands, but only the interact is thrown for interacting with things like sheep or villagers? (from what I remember 1.7-1.8 it was either or, never both, though I could be wrong).
- Possible names for the interacting at an entity is something like `InteractAtEntityPart` or something similar where the name is self describing that "hey, this event *will* have a local vector of where on the entity box was interacted by the player." Whereas the normal `InteractEntity` is sufficient for the original ported one that hasn't been touched in years.

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 7]
```diff
@@ -1,88 +1,322 @@
...
-    @Deprecated
-    public PlayerInteractEvent(EntityPlayer player, Action action, BlockPos pos, EnumFacing face, World world)
+        public Entity getTarget()
+        {
+            return target;
+        }
+    }
+
+    /**
+     * This event is fired on both sides after EntityInteractSpecific is fired above.
```
Comment (Aaron1011) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: I think saying 'after EntityInteractSpecific is fired` is slightly confusing. While the underlying `INTERACT_AT` packet is sent after the regular `INTERACT` one, there's no other relationship between the events. This sentence makes it sound as if the same underlying event directly triggers both.

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 8]
```diff
@@ -934,4 +935,44 @@ public static Vec3d rayTraceEyeHitVec(EntityLivingBase entity, double length)
         RayTraceResult git = rayTraceEyes(entity, length);
         return git == null ? null : git.hitVec;
     }
+
+    public static boolean onInteractEntityAt(EntityPlayer player, Entity entity, RayTraceResult ray, ItemStack stack, EnumHand hand)
```
Comment (kashike) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: These seem like they'd go better in `ForgeEventFactory`.

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 9]
```diff
@@ -1,88 +1,322 @@
...
+
+    /**
+     * This event is fired on both sides before the player triggers Item.onItemRightClick.
+     * This is the main item right click event, and called last out of ALL right click interactions, if the others have not caused the process to abort.
+     * Note that this is NOT fired if the player is targeting a block. For that case, see {@link RightClickBlock}.
+     * Canceling the event clientside causes processing to continue to the other hands,
+     * while canceling serverside will simply do no further processing.
+     */
+    @Cancelable
+    public static class ItemRightClick extends PlayerInteractEvent
```
Comment (kashike) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: `RightClickItem`

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 10]
```diff
@@ -1,88 +1,322 @@
...
+            super(player, hand, stack, new BlockPos(player), null);
+        }
+    }
+
+    /**
+     * This event is fired on the clientside when the player right clicks empty space with an empty hand.
+     * This event is fired last out of ALL right click interactions, if the others have not caused the process to abort.
+     * This event cannot be canceled.
+     */
+    public static class EmptyRightClick extends PlayerInteractEvent
```
Comment (kashike) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: `RightClickEmpty`

**Comment by williewillus** [comment]
@gabizou 
1. Both are fired for every entity you can target
2. Hmm maybe
3. The way its setup right now in vanilla makes sense. On the client, control simply flows all the way down and through all hands. Each part (left click block, right click block, etc.) send packets to the server. The server processes things in discrete chunks per packet. Canceling something on the server just makes it... *not happen*. Canceling something on the client causes flow to continue to whatever's next. That being said, canceling something on one side and not the other can cause desyncs

File: src/main/java/net/minecraftforge/common/ForgeHooks.java (hunk 8 above)
Comment (williewillus) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: No one has established a convention really for which goes where, I'm just going to leave them here until lex tells me to move them :P`

**Comment by gabizou** [comment]
> Each part (left click block, right click block, etc.) send packets to the server.

So am I right in understanding that the client will send a total of 2 packets when having a torch on the main hand and a shield on the offhand? One for an interact entity with the torch and one for a use item with the shield? (correct me on this one as I'm not at all sure the logic for it yet)?


> 1. Both are fired for every entity you can target

So now we have the interact at being fired first, and then, regardless of cancellation, the interact entity is thrown immediately after?

**Comment by williewillus** [comment]
1. send a total of 2 packets doing what?
2. yep

**Comment by gabizou** [comment]
> send a total of 2 packets doing what?

Send a total of 2 packets for simply right clicking with those items in the setup.

Re: InteractAt + InteractEntity: Which one of the packet handling in the `NetHandlerPlayServer` is actually calling which specific interact methods like `Item.onItemUse` or `Entity.interactWith(Player)` or however the method is structured?

**Comment by williewillus** [comment]
depends on what you're right clicking :P

I have traced the general flow of interaction logic in MC in a gist. Note that the flow described is how Forge 1.9 is, which means it doesn't describe the hooks I add here and some patches are glaringly missing like onItemuseFirst and doesSneakBypassUse that I have readded here
Gist: https://gist.github.com/williewillus/c781d9824df3ae723bff

**Comment by williewillus** [comment]
Updated to potentially clarify some things
Renamed ItemRightClick and EmptyRightClick to RightClickItem and RightClickEmpty

**Comment by Guichaguri** [comment]
> If i cancel a left click block i want to make it like it never happened

I found an issue: if you cancel left click server-side, the client-side tile entity is recreated without any data when the block is broken. This is a problem for signs or banners. It doesn't happen in `BreakEvent`.
Other than that, everything else seems to work properly

**Comment by williewillus** [comment]
Can you pastebin your test case so I can add it to the real test?

**Comment by Guichaguri** [comment]
@williewillus Create an event handler and add this code:
```java
    @SubscribeEvent
    public void leftClick(LeftClickBlock event) {
        if(event.getSide() == Side.SERVER) event.setCanceled(true);
    }
```
Here is a gif showing what happens: http://gfycat.com/OrneryLonelyGermanshorthairedpointer

**Comment by Guichaguri** [comment]
I also found another bug. This one is really annoying
Returning `true` in `Block.onBlockActivated` doesn't prevent the held item to be used.
No code is needed to reproduce this bug.
Another gif: http://gfycat.com/AcademicIdealisticGadwall

I've been using your fork for development for days and these are the just 2 bugs I found so far.

**Comment by williewillus** [comment]
^ Hm, I thought I fixed that one in later commits, make sure youre using the latest copy of my branch

**Comment by Guichaguri** [comment]
I'm using https://github.com/MinecraftForge/MinecraftForge/pull/2604/commits/f2a36e57333a60b4ac2092368de9ea603bbe9234. I'm going to update it anyway

**Comment by williewillus** [comment]
I think I found the issue, update later today maybe

**Comment by williewillus** [comment]
It was because I only patched the survival version of the call to onItemUse, not the creative one

**Comment by williewillus** [comment]
Okay, so when canceled serverside will now refresh the TE to the client using notifyBlockUpdate
Also added the patches for onItemUse to creative mode as well

**Comment by Guichaguri** [comment]
I can confirm that both of them are fixed. I'll keep an eye for other bugs.

**Comment by williewillus** [comment]
Squashed

**Comment by LexManos** [comment]
Well only way to get this going is to blame @williewillus and throw him under the bus for any bugs reported.
